pkg/controllers/provisioning/provisioner.go

```diff
@@ -372,14 +373,207 @@ func (p *Provisioner) getDaemonSetPods(ctx context.Context) ([]*v1.Pod, error) {
+	namespaceedLimitRanges, namespaceExists := limitRanges[pod.ObjectMeta.Namespace]
+	if !namespaceExists {
+		return true
+	}
+
+	// LimitRanges for Containers
+	for _, lr := range namespaceedLimitRanges {
+		for _, limitrangeItem := range lr.Spec.Limits {
+			if limitrangeItem.Type == v1.LimitTypePod {
+				if !applyLRTypePod(limitrangeItem, pod) {
+					return false
+				}
+			} else {
+				if !applyLRTypeContainer(limitrangeItem, pod) {
+					return false
+				}
+			}
+		}
+	}
+	return true
+}
+
+// Limit Ranges for Pods
+func applyLRTypePod(LRItem v1.LimitRangeItem, pod *v1.Pod) bool {
+	podTotalLimits := resources.LimitsForPods(pod)
+	podTotalRequests := resources.RequestsForPods(pod)
+	delete(podTotalLimits, "pods")
+	delete(podTotalRequests, "pods")
+
+	// Checks to see if the pod limits are bellow limitRange
+	// maximum resources allowed
+	if !resources.Fits(podTotalLimits, LRItem.Max) && LRItem.Max != nil {
+		return false
+	}
+	// Checks to see if the pod limits are above limitRange
+	// minimum resources allowed
+	if !resourceLRGreater(podTotalLimits, LRItem.Min) && LRItem.Min != nil {
+		return false
+	}
+	// Checks to see if the pod limits are within limitRange
+	// limit request ratio resources allowed
+	if !resourceLRRatio(podTotalLimits, podTotalRequests, LRItem.MaxLimitRequestRatio) && LRItem.MaxLimitRequestRatio != nil {
+		return false
+	}
+
+	return true
+}
+
+// Limit Ranges for Containers
+func applyLRTypeContainer(LRItem v1.LimitRangeItem, pod *v1.Pod) bool {
+	if !checkLRMax(LRItem.Max, pod) {
+		return false
+	}
+	if !checkLRMin(LRItem.Min, pod) {
+		return false
+	}
+	if !checkLRMaxLimitRequestRatio(LRItem.MaxLimitRequestRatio, pod) {
+		return false
+	}
+	applyLRDefault(LRItem.Default, pod)
+	applyLRDefaultRequest(LRItem.DefaultRequest, pod)
+
+	return true
+}
+
+func applyLRDefault(limitrangeDefault v1.ResourceList, pod *v1.Pod) {
+	if limitrangeDefault == nil {
+		return
+	}
+
+	for index := range pod.Spec.Containers {
+		limitResource := &pod.Spec.Containers[index].Resources.Limits
+		if *limitResource == nil {
+			*limitResource = limitrangeDefault.DeepCopy()
+		} else {
+			for resourceName := range limitrangeDefault {
+				_, exists := (*limitResource)[resourceName]
+				if !exists {
+					(*limitResource)[resourceName] = limitrangeDefault[resourceName].DeepCopy()
+				}
+			}
+		}
+	}
+}
+
+func applyLRDefaultRequest(limitrangeDefaultRequest v1.ResourceList, pod *v1.Pod) {
+	if limitrangeDefaultRequest == nil {
+		return
+	}
+
+	for index := range pod.Spec.Containers {
+		requestResource := &pod.Spec.Containers[index].Resources.Requests
+		if *requestResource == nil {
+			*requestResource = limitrangeDefaultRequest.DeepCopy()
+		} else {
+			for resourceName := range limitrangeDefaultRequest {
+				_, exists := (*requestResource)[resourceName]
+				if !exists {
+					(*requestResource)[resourceName] = limitrangeDefaultRequest[resourceName].DeepCopy()
```

Comment:
What occurs on a real cluster when there is a min request, and a default request , and the container request is too low? Does it just fail, or does it replace the lower request with the higher default one?

Reply:
After doing some test with the `kube-controller-manger` here is what I have found.

For Containers
- **Default**
- The values will be set for both limits and requests 
- If a daemonset resource limit is set  the daemonset resource request will be the same
- **defaultRequest** 
- The resources are set just for the daemonset resources requests

For Pods and Containers
- **Max**
- Checks to see if the limit is bellow the maximum resource list
- If the values of the daemonset limits resource list is above the limit range maximum resource list, the daemonset will not be scheduled 
- **Min**
- Checks to see if the requests is bellow the minimum resource list
- If the values of the daemonset requests resource list is bellow the limit range minimum resource list, the daemonset will not be scheduled 
- **maxLimitRequestRatio**
- All resources must be within a limil range resource ratio for all resources
- If any resources fall outside of that ratio, daemonset will not be scheduled